cou/apps/auxiliary.py

```diff
@@ -0,0 +1,180 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Auxiliary application class."""
+import logging
+from typing import Optional
+
+from cou.apps.app import AppFactory, OpenStackApplication
+from cou.exceptions import ApplicationError, HaltUpgradePlanGeneration
+from cou.steps import UpgradeStep
+from cou.utils.juju_utils import async_upgrade_charm
+from cou.utils.openstack import (
+    CHARM_TYPES,
+    LTS_SERIES,
+    AuxiliaryTrackMapping,
+    OpenStackRelease,
+)
+
+logger = logging.getLogger(__name__)
+
+
+@AppFactory.register_application(
+    ["rabbitmq-server", "vault"] + CHARM_TYPES["mysql"] + CHARM_TYPES["ovn"]
+)
+class AuxiliaryOpenStackApplication(OpenStackApplication):
+    """Application for charms that can have multiple OpenStack releases for a workload."""
+
+    openstack_to_track_mapping = AuxiliaryTrackMapping.generate_map()
+
+    def upgrade_description(self, target: OpenStackRelease) -> str:
+        """Top description for upgrading the application.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :return: Description of the update.
+        :rtype: str
+        """
+        return f"Upgrade plan for '{self.name}' to {target.codename}"
+
+    def openstack_to_track(self, os_release: OpenStackRelease) -> str:
+        """Find the track of auxiliary charms by Ubuntu release and OpenStack release codename.
+
+        :param os_release: OpenStack release to track.
+        :type os_release: OpenStackRelease
+        :raises ApplicationError: When there is no track compatible.
+        :return: The track of the auxiliary charm.
+        :rtype: str
+        """
+        try:
+            return self.openstack_to_track_mapping[self.series][self.charm][os_release.codename]
+        except KeyError as exc:
+            raise ApplicationError(
+                f"Not possible to find the track for '{self.charm}' on {os_release.codename}"
+            ) from exc
+
+    def os_origin_config(self, target: OpenStackRelease) -> Optional[OpenStackRelease]:
+        """Identify the OpenStack release set on openstack-origin or source config.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :return: OpenStackRelease object or None if the app doesn't have os_origin config.
+        :rtype: Optional[OpenStackRelease]
+        """
+        os_origin_parsed = None
+        if self.os_origin is not None:
```

Comment:
```
if !self.os_origin
    return ....
```

Reply:
Using `is not None` it's ok, but I agree that we could simple return the value, instead of defining `os_origin_parsed = None` and than do `OpenStackRelease(os_origin_parsed) if os_origin_parsed else None`.
Like this we are doing multiple if statements instead of single one.

Reply:
I refactored to use if else instead of match. However, we still need to keep `return OpenStackRelease(os_origin_parsed) if os_origin_parsed else None` because `target.previous_release` can be `None` and pylint will complain if we return direct